session/schema_amender.go

```diff
@@ -128,6 +130,72 @@ func (a *amendCollector) keyHasAmendOp(key []byte) bool {
 	return len(ops) > 0
 }
 
+func needCollectIndexOps(actionType uint64) bool {
+	return actionType&(1<<model.ActionAddIndex) != 0
+}
+
+func needCollectModifyColOps(actionType uint64) bool {
+	return actionType&(1<<model.ActionModifyColumn) != 0
+}
+
+func fieldTypeDeepEquals(ft1 *types.FieldType, ft2 *types.FieldType) bool {
+	if ft1.Tp == ft2.Tp &&
+		ft1.Flag == ft2.Flag &&
+		ft1.Flen == ft2.Flen &&
+		ft1.Decimal == ft2.Decimal &&
+		ft1.Charset == ft2.Charset &&
+		ft1.Collate == ft2.Collate &&
+		len(ft1.Elems) == len(ft2.Elems) {
+		for i, elem := range ft1.Elems {
+			if elem != ft2.Elems[i] {
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+// colChangeAmendable checks whether the column change is amendable, now only increasing column field
+// length is allowed for committing concurrent pessimistic transactions.
+func colChangeAmendable(colAtStart *model.ColumnInfo, colAtCommit *model.ColumnInfo) error {
+	// Modifying a stored generated column is not allowed by DDL, the generated related fields are not considered.
+	if !fieldTypeDeepEquals(&colAtStart.FieldType, &colAtCommit.FieldType) {
+		if colAtStart.FieldType.Flag != colAtCommit.FieldType.Flag {
+			return errors.Trace(errors.Errorf("flag is not matched for column=%v, from=%v to=%v",
+				colAtCommit.Name.String(), colAtStart.FieldType.Flag, colAtCommit.FieldType.Flag))
+		}
+		if colAtStart.Charset != colAtCommit.Charset || colAtStart.Collate != colAtCommit.Collate {
+			return errors.Trace(errors.Errorf("charset or collate is not matched for column=%v", colAtCommit.Name.String()))
+		}
+		err := ddl.CheckModifyTypeCompatible(&colAtStart.FieldType, &colAtCommit.FieldType)
+		if err != nil {
+			return errors.Trace(err)
+		}
```

Comment:
It seems that we need to consider `auto rand bit`.

Reply:
Will it be some problems if we commit using old `auto rand bit`? I'm not very sure.

Reply:
`auto-rand bit` operation is similar to `auto-increment id`. Refer to [auto-random](https://docs.pingcap.com/zh/tidb/stable/auto-random).